### PR TITLE
Azkaban to run user jobs using user credentials instead of using Azkaban

### PR DESCRIPTION
**Pull request by johnyu0520**
This will close a very big security hole where everybody who is running commands on the local Azkaban box is running as the user Azkaban.  In the not so bad case, different users might set Kerberos tickets and affect one another.  In the worst case, if Azkaban user has special privileges, people will be able to exploit that and wreak havoc

### CODE REVIEW

File: azkaban-execserver/src/main/resources/execute-as-user.c [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+#include <sys/stat.h>
+#include <sys/mount.h>
+#include <sys/types.h>
+#include <pwd.h>
+
+  FILE *LOGFILE = NULL;
+  FILE *ERRORFILE = NULL;
+  int SETUID_OPER_FAILED=10;
+
+/**
```
Comment (hluu) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: Would be good to mention that this code is borrowed from Hadoop

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java [hunk 2]
```diff
@@ -81,8 +83,23 @@ public void run() throws Exception {
...
     File[] propFiles = initPropsFiles();
     Map<String, String> envVars = getEnvironmentVariables();
-
+  
+    String nativeLibFolder = null;
+    String executeAsUserBinary = null;
+    String userToProxy = null;
+    boolean isExecuteAsUser = sysProps.getBoolean(EXECUTE_AS_USER, false);
+    
+    if(isExecuteAsUser){
```
Comment (hluu) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: Not exactly sure how we would go about doing unit testing on this feature, but something to think about.

File: azkaban-execserver/src/main/resources/execute-as-user.c [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        ERRORFILE=stderr;
+
+
+
+    char *uid = argv[1];
+
+    // for loop to calculate the length to malloc
+    int i;
+    int total_len = 0;
+    for(i=2;i<argc;i++){
```
Comment (logiclord) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: indenting is off for c file

File: azkaban-execserver/src/main/resources/execute-as-user.c [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        total_len += strlen(argv[i])+1;
+    }
+    printf("total_len: %d\n", total_len);
+
+    // allocate memory and clear memory
+    char *cmd = malloc(total_len+2);
+    memset(cmd, ' ', total_len+2);
```
Comment (logiclord) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: cmd can be NULL if system failed to allocate memory

Comment (logiclord) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: writing space will write int "32" in each byte for the memory allocated to char pointer. You can use cmd memset(cmd, '\0', total_len+2) or memset(cmd, 0, total_len+2)

File: azkaban-execserver/src/main/resources/execute-as-user.c [hunk 5]
```diff
@@ -0,0 +1,111 @@
+#include <dirent.h>
```
Comment (logiclord) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: license ?

File: azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java (hunk 2 above)
Comment (johnyu0520) [inline azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java]: I've confirmed the flag to work, which will enable the worst case scenario of rolling back.
However, how to warn users beforehand and the impact analysis will be something harder to do

File: azkaban-execserver/src/main/resources/execute-as-user.c (hunk 5 above)
Comment (johnyu0520) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: done

File: azkaban-execserver/src/main/resources/execute-as-user.c (hunk 1 above)
Comment (johnyu0520) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: done

File: azkaban-execserver/src/main/resources/execute-as-user.c (hunk 3 above)
Comment (johnyu0520) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: sorry, what should this be changed to?

File: azkaban-execserver/src/main/resources/execute-as-user.c (hunk 4 above)
Comment (johnyu0520) [inline azkaban-execserver/src/main/resources/execute-as-user.c]: fixed both malloc and memset

**Comment by johnyu0520** [comment]
closing the pull request for now as more modification is in progress

**Pull request closed without merging**